src/renderer/entities/price/model/price-provider-model.ts

```diff
@@ -0,0 +1,98 @@
+import { createEvent, createStore, forward, createEffect, sample } from 'effector';
+
+import { PriceApiProvider } from '../lib/types';
+import { DEFAULT_FIAT_PROVIDER, DEFAULT_ASSETS_PRICES, DEFAULT_FIAT_FLAG } from '../lib/constants';
+import { fiatService, coingekoService, PriceObject, PriceAdapter } from '@renderer/shared/api/price-provider';
+import { kernelModel } from '@renderer/shared/core';
+import * as currencyModel from './currency-model';
+import { chainsService } from '@renderer/entities/network';
+import { nonNullable } from '@renderer/shared/lib/utils';
+
+export const $fiatFlag = createStore<boolean | null>(null);
+export const $priceProvider = createStore<PriceApiProvider | null>(null);
+export const $assetsPrices = createStore<PriceObject | null>(null);
+
+const fiatFlagChanged = createEvent<boolean>();
+const priceProviderChanged = createEvent<PriceApiProvider>();
+const assetsPricesRequested = createEvent<boolean>();
+
+const getFiatFlagFx = createEffect((): boolean => {
+  return fiatService.getFiatFlag(DEFAULT_FIAT_FLAG);
+});
+
+const saveFiatFlagFx = createEffect((flag: boolean) => {
+  fiatService.saveFiatFlag(flag);
+});
+
+const getPriceProviderFx = createEffect((): PriceApiProvider => {
+  return fiatService.getFiatProvider(DEFAULT_FIAT_PROVIDER);
+});
+
+const savePriceProviderFx = createEffect((provider: string) => {
+  fiatService.saveFiatProvider(provider);
+});
+
+type FetchPrices = {
+  provider: PriceApiProvider;
+  currencies: string[];
+  includeRates: boolean;
+};
+const fetchAssetsPricesFx = createEffect<FetchPrices, PriceObject>(({ provider, currencies, includeRates }) => {
+  const ProvidersMap: Record<PriceApiProvider, PriceAdapter> = {
+    [PriceApiProvider.COINGEKO]: coingekoService,
+  };
+
+  const priceIds = chainsService.getChainsData().reduce<string[]>((acc, chain) => {
+    const ids = chain.assets.map((asset) => asset.priceId).filter(nonNullable);
+    acc.push(...ids);
+
+    return acc;
+  }, []);
+
+  return ProvidersMap[provider].getPrice(priceIds, currencies, includeRates);
+});
+
+const getAssetsPricesFx = createEffect((): PriceObject => {
+  return fiatService.getAssetsPrices(DEFAULT_ASSETS_PRICES);
+});
+
+const saveAssetsPricesFx = createEffect((prices: PriceObject) => {
+  fiatService.saveAssetsPrices(prices);
+});
+
+forward({
+  from: kernelModel.events.appStarted,
+  to: [getFiatFlagFx, getPriceProviderFx, getAssetsPricesFx],
+});
+
+forward({ from: getFiatFlagFx.doneData, to: $fiatFlag });
+
+forward({ from: getPriceProviderFx.doneData, to: $priceProvider });
+
+forward({ from: getAssetsPricesFx.doneData, to: $assetsPrices });
+
+sample({
+  clock: [$priceProvider, currencyModel.$activeCurrency],
```

Comment:
is clock really needed here? sample will be run on source changed, and here source is the same as clock

Reply:
Added missing `assetsPricesRequested` to clock array and it will need data from the source
Otherwise yes, you're right about source update without clock 👍

```typescript
sample({
  clock: [assetsPricesRequested, $priceProvider, currencyModel.$activeCurrency],
  source: { provider: $priceProvider, currency: currencyModel.$activeCurrency },
  filter: ({ provider, currency }) => provider !== null && currency !== null,
  fn: ({ provider, currency }) => {
    return { provider: provider!, currencies: [currency!.coingeckoId], includeRates: true };
  },
  target: fetchAssetsPricesFx,
});
```

Reply:
By the way I've asked about type guards for complex arguments, and core contributor told me not to hesitate using bang operator, because we are sure in our filter
Or we should provide a type guard with full typing which is really verbose and seems bizzare

> Подскажите, как корректно следить за 2 сторами в sample, сторы могут быть null, 
> и filter должен пропускать только, если оба стора не пусты?
> 
> Получается что-то вроде этого, но TS не пускает из-за error: 'fn result should extend target type';
> 
> sample({
>   clock: [$first, $second],
>   source: { first: $first, second: $second },
>   filter: ({ first, second }) => first && second,
>   fn: ({ first, second }) => ({ one: first, two: second }),
>   target: myEffect,
> });
----

> всё правильно, только в filter сделай корректную проверку на null, а то сейчас 0 и пустую строку не пропустит тоже.
> чтобы успокоить ts можно или типизировать filter для сужения типа source, или проще поставить восклицательные знаки в fn.

> если нет предубеждений против !, то я бы просто их повтыкал :) так ты просто говоришь ts, что тут точно не будет null|undefined, что ты лучше знаешь